perf: Cache search options

TMisiukiewicz opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR introduces performance improvement for loading searchable lists by caching the options and filtering them when accessing the list.
### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/37878
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Before each test: close & open or refresh the app, to make sure options are not cached before first opening

**Search page:**
1. Open the search page
2. Verify the list loaded
3. Close the list
4. Open it again and make sure it loaded

**New chat**
1. Press "Start chat"
2. Verify the list loaded - it should contain user name/email/phone number
3. Close the list
4. Open it again and make sure it loaded

**New room chat**
1. Press "Start chat"
2. Select "Room"
3. Fill the name and create a room
4. Open room details
5. Press "Members"
6. Press "Invite"
7. Verity the list loaded with user name/email/phone number
8. Go back
9. Press "Invite" again
10. Verify the list loaded

**Request money**
1. Press "Request money"
2. Select "Manual", fill the amount and press "Next"
3. Verify the list loaded
4. Go back and press "Next" again
5. Verify the list loaded
6. Select an item to split bill and finish the process of splitting

**Send money**
1. Press "Send money"
2.  Fill the amount and press "Next"
3. Verify the list loaded
4. Go back and press "Next again"
5. Verify the list loaded

**Assign task**
1. Press "Assign task"
2. Fill the title and press "Next"
3. Press "Assignee"
4. Verify the list loaded with user name/email/phone number
5. Go back and press "Assignee" again
6. Verify the list loaded

**Assign task - share somewhere**
1. Press "Assign task"
2. Fill the title and press "Next"
3. Press "Share somewhere"
4. Verify the list loaded with reports
5. Go back and press "Share somewhere" again
6. Verify the list loaded

**Invite to workspace**
1. Open "Choose a workspace"
2. Press "+" button
3. Go to "Members"
4. Press "Invite member"
5. Verify the list loaded with user name/email/phone number
6. Go back and press "Invite member" again
7. Verify the list loaded again
8. Select an user and press "Next"
9. Verify you were moved to another step

**Share logs**
1. Go to Settings -> About -> Troubleshoot
2. Enable "Client Side logging"
3. Press "Debug console"
4. Press "Share log"
5. Verify the list is visible
6. Go back and press "Share log" again
7. Verify the list is visible 
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
12. Upload an image via copy paste
13. Verify a modal appears displaying a preview of that image
--->
Before each test: close & open or refresh the app, to make sure options are not cached before first opening

**Search page:**
1. Open the search page
2. Verify the list loaded
3. Close the list
4. Open it again and make sure it loaded

**New chat**
1. Press "Start chat"
2. Verify the list loaded - it should contain user name/email/phone number
3. Close the list
4. Open it again and make sure it loaded

**New room chat**
1. Press "Start chat"
2. Select "Room"
3. Fill the name and create a room
4. Open room details
5. Press "Members"
6. Press "Invite"
7. Verity the list loaded with user name/email/phone number
8. Go back
9. Press "Invite" again
10. Verify the list loaded

**Request money**
1. Press "Request money"
2. Select "Manual", fill the amount and press "Next"
3. Verify the list loaded
4. Go back and press "Next" again
5. Verify the list loaded
6. Select an item to split bill and finish the process of splitting

**Send money**
1. Press "Send money"
2.  Fill the amount and press "Next"
3. Verify the list loaded
4. Go back and press "Next again"
5. Verify the list loaded

**Assign task**
1. Press "Assign task"
2. Fill the title and press "Next"
3. Press "Assignee"
4. Verify the list loaded with user name/email/phone number
5. Go back and press "Assignee" again
6. Verify the list loaded

**Assign task - share somewhere**
1. Press "Assign task"
2. Fill the title and press "Next"
3. Press "Share somewhere"
4. Verify the list loaded with reports
5. Go back and press "Share somewhere" again
6. Verify the list loaded

**Invite to workspace**
1. Open "Choose a workspace"
2. Press "+" button
3. Go to "Members"
4. Press "Invite member"
5. Verify the list loaded with user name/email/phone number
6. Go back and press "Invite member" again
7. Verify the list loaded again
8. Select an user and press "Next"
9. Verify you were moved to another step

**Share logs**
1. Go to Settings -> About -> Troubleshoot
2. Enable "Client Side logging"
3. Press "Debug console"
4. Press "Share log"
5. Verify the list is visible
6. Go back and press "Share log" again
7. Verify the list is visible 
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/13985840/ca517bc9-98fd-4a35-97f9-39bd3c6785f8


https://github.com/Expensify/App/assets/13985840/40f97f3e-439e-4b7f-b3ff-5d0eddf5c0af


https://github.com/Expensify/App/assets/13985840/ba71cccb-9552-411b-b197-0991adbab947


https://github.com/Expensify/App/assets/13985840/f8de3bfe-a9c0-4f9c-a6a1-8b6491cb26f9


https://github.com/Expensify/App/assets/13985840/cfe891ae-1d58-4087-b100-415424877b48


https://github.com/Expensify/App/assets/13985840/4f802a78-190d-49bd-8381-4ceb56531cf9


https://github.com/Expensify/App/assets/13985840/c5c3ad71-6bee-48ae-af99-f3587eb366ba


https://github.com/Expensify/App/assets/13985840/b861a592-d52d-49e8-b747-4266df4d0c61


https://github.com/Expensify/App/assets/13985840/a12d9483-3219-4313-bc7f-5d438657f9b6




<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/13985840/22713558-a669-45a4-9079-b8f35d40bbc1


https://github.com/Expensify/App/assets/13985840/f93a7de1-1dc8-4dba-9b78-cfe8149abc41


https://github.com/Expensify/App/assets/13985840/9c4da3f5-8ad2-41ee-b154-e2ff3af5bca7


https://github.com/Expensify/App/assets/13985840/09663ff6-7c06-4c1f-9983-44ed536a9fc9


https://github.com/Expensify/App/assets/13985840/97f71aa7-55c1-4c5e-adbe-a09d5c521908


https://github.com/Expensify/App/assets/13985840/2aba17ac-ae5d-40b7-945c-c6a5d5521781


https://github.com/Expensify/App/assets/13985840/8b88c935-46ca-43c0-9cf1-54ffb8e964be


https://github.com/Expensify/App/assets/13985840/4fa8d55c-39ec-40a8-9487-8ca6ee7cb23f


https://github.com/Expensify/App/assets/13985840/20abdeb4-3c7c-4be6-8228-80a794cd91c9


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/13985840/5bca863e-cb4f-4d0f-aa75-93f8435a6281


https://github.com/Expensify/App/assets/13985840/4bc076ec-09be-460b-86fc-590f4c046563


https://github.com/Expensify/App/assets/13985840/ae6f36c8-f797-447e-ae15-209c46bbc11c


https://github.com/Expensify/App/assets/13985840/951f4f8c-23bd-45f5-98c8-4234f0a71ae7


https://github.com/Expensify/App/assets/13985840/d6970e36-21c9-4c17-8aea-e21642f9271f


https://github.com/Expensify/App/assets/13985840/8e2ef8fc-dfde-4443-ae57-c0946dbab44e


https://github.com/Expensify/App/assets/13985840/1600d90c-55e3-4e5d-9d60-c91871feccc7


https://github.com/Expensify/App/assets/13985840/e73d71f6-85fa-4469-baa8-8d5988e2bd70



<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/13985840/0f5b0204-c937-4cab-9bb7-bb2f5f655793


https://github.com/Expensify/App/assets/13985840/9cb5b741-3429-41bd-86c7-62c3f6d8cef4


https://github.com/Expensify/App/assets/13985840/23d6460b-8c83-406a-add5-b9b2f8bff98a


https://github.com/Expensify/App/assets/13985840/cf16082a-3caf-4225-8949-0aab00a817c6


https://github.com/Expensify/App/assets/13985840/7b4e9f18-bd18-482a-a5d5-10700d2eef54


https://github.com/Expensify/App/assets/13985840/527eae02-8f20-4c11-942f-93c684ba9c1e


https://github.com/Expensify/App/assets/13985840/e2269537-031b-462e-bdbe-7742099ce25f


https://github.com/Expensify/App/assets/13985840/db7cffcb-6b6e-4e90-99f3-bb7214fbd698


https://github.com/Expensify/App/assets/13985840/2d230c10-bc2d-468c-a498-fcbab33b17a5


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/13985840/3440c0ad-8041-4c9b-994d-8267348005a8


https://github.com/Expensify/App/assets/13985840/bcff5050-2665-4cd8-a70c-1a239252b690


https://github.com/Expensify/App/assets/13985840/aa6d75c7-6346-4706-8024-36c297671a78


https://github.com/Expensify/App/assets/13985840/97eca517-b235-4fb2-9d19-4cafd7e209d2


https://github.com/Expensify/App/assets/13985840/04daa5e7-78ce-4d2b-8429-9b23c83c2059


https://github.com/Expensify/App/assets/13985840/7807ca0e-5ebd-4cce-a3f6-8bea45791e2f


https://github.com/Expensify/App/assets/13985840/e24cbf85-046d-459f-b0c0-70912a352e97


https://github.com/Expensify/App/assets/13985840/ecc4e174-5f9d-4a4e-a73b-34418831a2ee



<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/13985840/1f9da805-2123-4d05-afd5-a0d7307a84ba


https://github.com/Expensify/App/assets/13985840/c294d986-4fb4-4454-b811-96763faba845


https://github.com/Expensify/App/assets/13985840/abc03d7f-80e9-4338-b2bf-9e6261a2f6b6


https://github.com/Expensify/App/assets/13985840/99759474-87cf-4fb1-ba0a-cdf1263a77bc


https://github.com/Expensify/App/assets/13985840/5394effd-fcb1-44eb-a78e-c1957327f155


https://github.com/Expensify/App/assets/13985840/f2ef66e0-5634-4cde-bc86-e3f9119b2df3


https://github.com/Expensify/App/assets/13985840/bd33f17f-a217-4e93-8c69-4d008ee0cfd0


https://github.com/Expensify/App/assets/13985840/aca064fc-851f-4172-a74b-10011cab4fd2


https://github.com/Expensify/App/assets/13985840/acb14e98-ff51-47de-a128-5623d38cbdb9




<!-- add screenshots or videos here -->

</details>

TMisiukiewicz committed:
add function to create all reports and personal details options

TMisiukiewicz committed:
use context to initialize all options

TMisiukiewicz committed:
update getOptions, use options generation in search

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

TMisiukiewicz committed:
update usage in search page

TMisiukiewicz committed:
update options list utils

TMisiukiewicz committed:
create reusable initializer for search pages

TMisiukiewicz committed:
add options to new chat page

TMisiukiewicz committed:
update other occurencies of getoptions

TMisiukiewicz committed:
update initializing in search page

TMisiukiewicz committed:
update types for options

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

TMisiukiewicz committed:
update loading in search page

TMisiukiewicz committed:
update new chat page skeleton displaying

TMisiukiewicz committed:
update caching in all selection lists

TMisiukiewicz committed:
update options in the background

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/cache-options

TMisiukiewicz committed:
update changed personal details

TMisiukiewicz committed:
fix loader on new chat page

TMisiukiewicz committed:
filter personal details with login

TMisiukiewicz committed:
generate alternate text on the fly if needed

TMisiukiewicz committed:
freeze the option list when search is open

TMisiukiewicz committed:
remove unnecessary code

TMisiukiewicz committed:
remove console logs

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

TMisiukiewicz committed:
fix for generating alternate text

TMisiukiewicz committed:
fix jest tests

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/cache-options

TMisiukiewicz committed:
update reassure tests

TMisiukiewicz committed:
fix prettier

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

TMisiukiewicz committed:
update reassure option list utils tests

TMisiukiewicz committed:
remove redundant function calls

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

TMisiukiewicz committed:
speed up setting options

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

TMisiukiewicz committed:
update money request participants selector

TMisiukiewicz committed:
lint code

TMisiukiewicz committed:
add comments for option list context

mountiny commented:
Discussing [here](https://expensify.slack.com/archives/C05LX9D6E07/p1710951193656849?thread_ts=1710942725.509289&cid=C05LX9D6E07)

TMisiukiewicz committed:
add test for cached options in search page

TMisiukiewicz committed:
update displaying lists

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

hungvu193 commented on src/components/OptionListContextProvider.tsx:
```diff
@@ -0,0 +1,137 @@
+        const newOption = OptionsListUtils.createOptionFromReport(lastUpdatedReport, personalDetails);
+        const replaceIndex = options.reports.findIndex((option) => option.reportID === lastUpdatedReport.reportID);
+
+        if (replaceIndex === undefined) {
```
isn't that this won't be never passed because `findIndex` will return index of the first element or -1.

hungvu193 commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

hungvu193 commented on src/libs/ReportUtils.ts:
```diff
@@ -512,6 +512,23 @@ Onyx.connect({
+    },
+});
+
+function getLastUpdatedReport() {
```
Please add the return value type for this function.

hungvu193 commented on src/pages/NewChatPage.tsx:
```diff
@@ -262,7 +264,7 @@ function NewChatPage({betas, isGroupChat, personalDetails, reports, isSearchingF
                             boldStyle
                             shouldPreventDefaultFocusOnSelectRow={!DeviceCapabilities.canUseTouchScreen()}
-                            shouldShowOptions={isOptionsDataReady && didScreenTransitionEnd}
+                            shouldShowOptions={areOptionsInitialized && didScreenTransitionEnd}
```
```suggestion
                            shouldShowOptions={areOptionsInitialized}
```
This will be enough right, because we have this above:
```javascript
    const {options, areOptionsInitialized} = useOptionsList({
        shouldInitialize: didScreenTransitionEnd,
    });
```

and inside our `useOptionsList`, we won't `initializeOptions()` if below conditions passed:
```
        if (!shouldInitialize || areOptionsInitialized) {
            return;
        }
```

hungvu193 commented on src/pages/tasks/NewTaskPage.tsx:
```diff
@@ -178,7 +178,7 @@ function NewTaskPage({task, reports, personalDetails}: NewTaskPageProps) {
                                 icon={shareDestination?.icons}
                                 onPress={() => Navigation.navigate(ROUTES.NEW_TASK_SHARE_DESTINATION)}
-                                interactive={!task?.parentReportID}
+                                interactive
```
I'm a little bit outdated about NewTaskPage, but can you explain why `interactive` will be always true here?

TMisiukiewicz replied:
good catch, it was mistakenly pushed, thanks!

hungvu193 commented on src/pages/tasks/TaskAssigneeSelectorModal.tsx:
```diff
@@ -220,14 +217,14 @@ function TaskAssigneeSelectorModal({reports, task}: TaskAssigneeSelectorModalPro
                     <View style={[styles.flex1, styles.w100, styles.pRelative]}>
                         <SelectionList
-                            sections={didScreenTransitionEnd && !isLoading ? sections : []}
+                            sections={(!areOptionsInitialized && didScreenTransitionEnd) || areOptionsInitialized ? sections : []}
```
I think we can do it similar to `NewChatPage` as I suggested above

```suggestion
                            sections={areOptionsInitialized ? sections : []}
```
and update `useOptionsList`:

```javascript
                             const {options, areOptionsInitialized} = useOptionsList({
        shouldInitialize: didScreenTransitionEnd,
    });
```

hungvu193 commented on src/pages/tasks/TaskAssigneeSelectorModal.tsx:
```diff
@@ -220,14 +217,14 @@ function TaskAssigneeSelectorModal({reports, task}: TaskAssigneeSelectorModalPro
                             headerMessage={headerMessage}
                             textInputLabel={translate('optionsSelector.nameEmailOrPhoneNumber')}
-                            showLoadingPlaceholder={isLoading || !didScreenTransitionEnd}
+                            showLoadingPlaceholder={!areOptionsInitialized && !didScreenTransitionEnd}
```
```suggestion
                            showLoadingPlaceholder={!areOptionsInitialized}
```

hungvu193 commented on src/pages/tasks/TaskShareDestinationSelectorModal.tsx:
```diff
@@ -104,13 +108,13 @@ function TaskShareDestinationSelectorModal({reports, isSearchingForReports}: Tas
                         <SelectionList
                             ListItem={UserListItem}
-                            sections={didScreenTransitionEnd ? options.sections : []}
+                            sections={(!areOptionsInitialized && didScreenTransitionEnd) || areOptionsInitialized ? options.sections : []}
```
Same with `TaskAssigneeSelectorModal`

hungvu193 reviewed (commented):
@TMisiukiewicz I left few comments, can we discuss about them? Thank you 🙇

hungvu193 commented:
Inside the `Tests` and `QA Steps`, you declared tests for all the flows, but I only saw the search part on your checklist videos. Can you please update the QA and Tests steps or videos?.  Also these flows are failing for me:

https://github.com/Expensify/App/assets/16502320/088a8d27-da62-4f0a-b0aa-9468310b7fdf


https://github.com/Expensify/App/assets/16502320/59766ade-d93a-4cf8-8b03-031896eef507


https://github.com/Expensify/App/assets/16502320/6f3e679c-1f83-4731-be5e-e01e2f259347

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/cache-options

TMisiukiewicz committed:
code review updates

TMisiukiewicz committed:
fix displaying personal details in current user option

TMisiukiewicz commented:
@hungvu193 thanks for your review, I already addressed all the comments and fixed one issue i stumbled upon while testing it. I am working on preparing videos for all scenarios.

Could you elaborate what exactly is failing in the flows on the videos you posted?

hungvu193 commented:
> @hungvu193 thanks for your review, I already addressed all the comments and fixed one issue i stumbled upon while testing it. I am working on preparing videos for all scenarios.
> 
> 
> 
> Could you elaborate what exactly is failing in the flows on the videos you posted?

I saw the skeleton view every time I opened these pages, I thought it was cached after the first time. Is that true?

TMisiukiewicz committed:
update caching behavior in lists

hungvu193 commented:
@TMisiukiewicz let me know when it's ready 😄. Thank you

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/cache-options

TMisiukiewicz commented:
@hungvu193 I've updated the videos section with videos for each flow 👍 everything should be ready now

hungvu193 commented:
I'm getting this crash when I tried to run, also please resolve the conflict 😄 

https://github.com/Expensify/App/assets/16502320/64c22772-0e9d-4bab-acfa-f9025b86c984

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

TMisiukiewicz committed:
update tagpicker types

TMisiukiewicz commented:
@hungvu193 Conflict resolved ✅  I believe this error comes from main, I was having it as well from time to time but couldn't reproduce after merging latest main. I see it was reported in #38972

hungvu193 commented:
> @hungvu193 Conflict resolved ✅  I believe this error comes from main, I was having it as well from time to time but couldn't reproduce after merging latest main. I see it was reported in #38972

Awesome, let me check it

hungvu193 reviewed (approved):
Test well!

hungvu193 commented on src/libs/ReportUtils.ts:
```diff
@@ -514,6 +514,23 @@ Onyx.connect({
+let lastUpdatedReport: OnyxEntry<Report>;
+
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.REPORT,
```
NAB but I think we're optimizing and use `waitForCollectionCallback: true` for all the report collections [here](https://github.com/Expensify/App/pull/38464), so I think it's good to has it here as well but if this PR can be merged before #38464 then I think I'm fine with this 😄

grgia replied:
should we address this given the linked PR was merged? @hungvu193

hungvu193 replied:
I think Yes, cc @TMisiukiewicz

TMisiukiewicz replied:
sorry I missed that comment before. I've tried it and seems like it does not work properly with `waitForCollectionCallback`, I am not getting an update of the item here when it's set to `true`

hungvu193 replied:
hmm yeah, I also got the issue, let skip it for now.

hungvu193 reviewed (approved):
@TMisiukiewicz Can you merge main so we can pass the Perf test? Thank you.
All yours @mountiny

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/cache-options

TMisiukiewicz commented:
@hungvu193 thanks, updated with the latest main. Just a reminder that Reassure tests are expected to fail here, detailed description is in [this comment](https://github.com/Expensify/App/pull/38207#issuecomment-2009918257)

TMisiukiewicz committed:
fix failing reassure tests

TMisiukiewicz committed:
Merge branch 'main' into perf/cache-options

TMisiukiewicz committed:
lint code

TMisiukiewicz committed:
fix typecheck

hungvu193 commented:
friendly bump @grgia

hungvu193 commented:
We have conflict again @TMisiukiewicz

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/cache-options

TMisiukiewicz commented:
@hungvu193 thanks, resolved ✅

grgia commented:
Triggering a test build

grgia commented:
Rerunning desktop

grgia commented on src/libs/OptionsListUtils.ts:
```diff
@@ -509,6 +518,28 @@ function getLastActorDisplayName(lastActorDetails: Partial<PersonalDetails> | nu
         : '';
 }
 
+/**
+ * Update alternate text option when applicable
+ */
+function getAlternateTextOption(
```
What's the alternative text option?

TMisiukiewicz replied:
right, wrong naming here. Updated it with `getAlternateText`

grgia commented on src/libs/OptionsListUtils.ts:
```diff
@@ -1320,12 +1350,63 @@ function isReportSelected(reportOption: ReportUtils.OptionData, selectedOptions:
+            }
+
+            const isSelfDM = ReportUtils.isSelfDM(report);
+            // Currently, currentUser is not included in visibleChatMemberAccountIDs, so for selfDM we need to add the currentUser as participants.
```
are we planning on changing this? In what case do we care about the selfDM accountID

TMisiukiewicz replied:
Hard to say, those changes were originally made in #37165, I just extracted it to separate function

grgia reviewed (dismissed):
LGTM and tests well, a few NAB comments

TMisiukiewicz committed:
update function naming

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/cache-options

TMisiukiewicz commented:
Resolved conflicts and addressed the comments.

grgia reviewed (approved):
Nice work on this

grgia commented:
See: https://github.com/Expensify/App/pull/38207#issuecomment-2022668486 for why the merge with failing test

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

bondydaa commented:
@mountiny was this PR actually deployed? I see you linked in the Deploy Checklist but that last comment here confuses me.

mountiny commented:
@bondydaa Yes it was deployed, the github action to post comment has also been broken 😵‍💫

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8598730206) to production by https://github.com/Julesssss in version: 1.4.60-13 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

jjcoffee commented on src/components/OptionListContextProvider.tsx:
```diff
@@ -0,0 +1,142 @@
+    });
+    const personalDetails = usePersonalDetails();
+
+    useEffect(() => {
```
Coming from https://github.com/Expensify/App/issues/40493, we missed updating the cache when a report is deleted.
